Linked apps: incremental versioning

orangejenny opened the pull request:
https://dimagi-dev.atlassian.net/browse/ICDS-612

Opening for review/discussion/tests. This is a small part of the [multiple master apps spec](https://docs.google.com/document/d/1KlY7LfMaQa0efGl8S58JQ9fDJ0JNj1eeTeB8aYvjg_4/edit#). I'm working on it because it would get us a step closer to being able to revert linked apps, which is high priority for ICDS (https://dimagi-dev.atlassian.net/browse/ICDS-552).

My intent is to build off of this to address the issues with reverting linked apps and also for this to fit into @kaapstorm 's work on the rest of the multiple master spec.

Feature flag: linked project spaces

orangejenny committed:
Added pulled_from_master_version to LinkedApplication

orangejenny committed:
Added pulled_from_master_version to saved_app view

orangejenny committed:
Added UI indicator for pulled_from_master_version to releases table

orangejenny committed:
Added pulled_from_master_app_version to ES

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -5545,6 +5545,8 @@ class LinkedApplication(Application):
     # This is the id of the master application
     master = StringProperty()
 
+    pulled_from_master_version = IntegerProperty()
```
nit: how about `pulled_master_version` or just `master_version`?

proteusvacuum replied:
I like `pulled_from_master_version`. It tells you exactly what it is.

orangejenny replied:
Keeping the original, explicit name.

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -5579,6 +5581,14 @@ def get_latest_master_release(self):
+    def wrap(cls, data):
+        self = super(LinkedApplication, cls).wrap(data)
+        if not self.pulled_from_master_version:
+            self.pulled_from_master_version = self.version
```
i think we should add a note why this is done?
my assumption is for linked applications before this change.

orangejenny replied:
sure - ebd39150e22

mkangia commented on corehq/apps/app_manager/templates/app_manager/partials/settings/app_settings.html:
```diff
@@ -194,7 +194,10 @@ <h6>Example:</h6>
+              <p>
+                {% blocktrans with version=app.pulled_from_master_version %}
+                  Your app was last pulled from master version {{ version }}.
+                {% endblocktrans %}
```
Is the order of things better like
```
Your app is at version 123. The master app is at version 456.
Your app was last pulled from master version 123
```

i am guessing the first and third statements are going to always have the same number for now but
1. The user might not always be able to connect that the latest pulled version is the current version
2. the spec said that the version of master and pulled are going to be different in future? so this would then become relevant.

kaapstorm replied:
If I understand correctly (and please correct me if I'm wrong) the purpose of this PR is to disconnect app version from master app version. This would allow us to revert linked apps independently from whatever is happening with the master app. So that example might be more like,

> Your app is at version 2. The master app is at version 456. Your app was last pulled from master version 123.

So it might be confusing to associate app version with master app version in the user's mental model, and I think maybe better to leave out the app version in this context. But again, please let me know if I've misunderstood the context or the approach that this PR is taking.

orangejenny replied:
@kaapstorm 's description and example are right. I think the point of this statement it to tell you whether or not the linked app is up to date with the master app, to help you decide if you should pull the master. To decide that, it doesn't matter what version the linked app is, and I'd rather not introduce one more number.

mkangia commented on corehq/apps/app_manager/views/utils.py:
```diff
@@ -346,7 +346,9 @@ def update_linked_app(app, user_id):
-            app = overwrite_app(app, latest_master_build, report_map)
+            new_version = app.version + 1 if app.version else 1
+            app = overwrite_app(app, latest_master_build, report_map, version=new_version)
+            app.pulled_from_master_version = latest_master_build.version
```
nit:out of scope of this PR but a small change,
`latest_master_build` should be `latest_master_releaed_build`

snopoke replied:
you mean renaming the var? `latest_master_build` is already the latest released master build in reality

mkangia replied:
ya i just meant the var name.

orangejenny replied:
87a33b077d8

mkangia commented on corehq/apps/app_manager/templates/app_manager/partials/releases/releases_table.html:
```diff
@@ -160,7 +160,11 @@
-          <strong>{% trans "Version" %} <span data-bind="text: version"></span></strong> |
+          <strong>{% trans "Version" %} <span data-bind="text: version"></span></strong>
+          <span data-bind="visible: doc_type() == 'LinkedApplication'">
+            (master v<span data-bind="text: pulled_from_master_version"></span>)
```
i am not sure if space is a constraint.
this would be clearer with `pulled from master v` instead.

orangejenny replied:
Leaving this as-is, because of space. There's going to be even more info crammed in here when it also needs to note which master app was pulled.

mkangia reviewed (approved):
Looks great. nothing blocking.

mkangia commented:
@orangejenny 
do you need to mention the FF this is behind in the description since the PR has that label?

snopoke commented on corehq/pillows/mappings/app_mapping.py:
```diff
@@ -4,7 +4,7 @@
 from pillowtop.es_utils import ElasticsearchIndexInfo
 
-APP_INDEX = es_index("hqapps_2019-01-23")
+APP_INDEX = es_index("hqapps_2019-05-17")
```
I don't think it's necessary to change the index name - you could just update the mapping since you're just adding a new field. You can update the mapping in a Django migration like this: https://github.com/dimagi/commcare-hq/blob/a13f0e173395a5415df5fe6d06220cfe4d0cabff/corehq/apps/case_search/migrations/0008_auto_20180119_1716.py#L25

orangejenny replied:
It looks like that migration was auto-generated, but how? `makemigrations` seems to only deal with sql models. When I try writing the migiration by hand I run into errors about conflicting migrations and attempting to merge them throws a "Could not find common ancestor". The migration looks like this (plus imports):

```
def update_es_mapping(*args, **kwargs):
    def _update_es_mapping():
        return call_command('update_es_mapping', 'app', noinput=True)
    return _update_es_mapping

class Migration(migrations.Migration):
    operations = [
        migrations.RunPython(update_es_mapping)
    ]
```

proteusvacuum replied:
You need a `dependencies` section in the `Migration` class. This can be the name of the previous migration in the same app. 

https://github.com/dimagi/commcare-hq/blob/a13f0e173395a5415df5fe6d06220cfe4d0cabff/corehq/apps/case_search/migrations/0008_auto_20180119_1716.py#L20-L22

orangejenny replied:
thanks - e06d064d7df

orangejenny committed:
Added test

orangejenny committed:
Added comment

orangejenny committed:
Renamed variable

orangejenny committed:
Added migration for ES index update, removed index renaming

mkangia commented:
@orangejenny 
build failure looks related
```
======================================================================
FAIL: corehq.apps.linked_domain.tests.test_linked_apps:TestLinkedApps.test_incremental_versioning
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/linked_domain/tests/test_linked_apps.py", line 142, in test_incremental_versioning
    self.assertEqual(current_master.version, 5)
AssertionError: 7 != 5
-------------------- >> begin captured logging << --------------------
celery.app.trace: INFO: Task corehq.apps.app_manager.tasks.prune_auto_generated_builds[22254f65-84ef-4e86-a333-18815ad25a30] succeeded in 0.0381024230001s: None
notify: ERROR: Notify Exception: Error updating CallCenter config for app build
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/signals.py", line 30, in update_callcenter_config
    cc_config = domain_obj.call_center_config
AttributeError: 'NoneType' object has no attribute 'call_center_config'
celery.app.trace: INFO: Task corehq.apps.app_manager.tasks.prune_auto_generated_builds[8c686a4d-affa-4090-a542-3bada311dc40] succeeded in 0.0316827009999s: None
notify: ERROR: Notify Exception: Error updating CallCenter config for app build
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/signals.py", line 30, in update_callcenter_config
    cc_config = domain_obj.call_center_config
AttributeError: 'NoneType' object has no attribute 'call_center_config'
celery.app.trace: INFO: Task corehq.apps.app_manager.tasks.prune_auto_generated_builds[2eb3b0bd-0f89-4773-bbd1-95d5bba6f3b3] succeeded in 0.0367284210001s: None
notify: ERROR: Notify Exception: Error updating CallCenter config for app build
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/signals.py", line 30, in update_callcenter_config
    cc_config = domain_obj.call_center_config
AttributeError: 'NoneType' object has no attribute 'call_center_config'
celery.app.trace: INFO: Task corehq.apps.app_manager.tasks.prune_auto_generated_builds[b7ffe132-a70a-49dc-a659-46ac3ef635ad] succeeded in 0.0325854990001s: None
notify: ERROR: Notify Exception: Error updating CallCenter config for app build
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/signals.py", line 30, in update_callcenter_config
    cc_config = domain_obj.call_center_config
AttributeError: 'NoneType' object has no attribute 'call_center_config'
```

orangejenny committed:
Updated new test to be more flexible when running with other tests

orangejenny committed:
Added default logic for pulled_from_master_version in couch view

orangejenny commented:
Added test, manually tested on staging. Clayton is checking that this doesn't interfere with ICDS's current stream hotfixing process, and then this should be good to release.

calellowitz commented on corehq/apps/app_manager/models.py:
```diff
@@ -5579,6 +5581,17 @@ def get_latest_master_release(self):
 
+    @classmethod
+    def wrap(cls, data):
+        self = super(LinkedApplication, cls).wrap(data)
```
using `self` as a local variable in this method is a little confusing but I see we do it elsewhere in the file, so if you prefer it, it seems fine to keep.

Most of the lazy migrations we do seem to modify `data`, then return `super` on the modified dict, instead of modifying the returned object. Is there a reason to prefer one or the other here?

orangejenny replied:
fe44da88d5d

orangejenny committed:
Updated LinkedApplication.wrap to be more like other wraps

mkangia commented:
@orangejenny 
how does this effect media versions? 
worried if this would not work well with 
https://github.com/dimagi/commcare-hq/pull/24285

orangejenny committed:
Updated application saving to increment version even for linked apps.

This allows linked apps to create new builds after changing
independent/overridable settings like CommCare flavor.

orangejenny commented:
Closing - this is going to be combined with other multi master changes in a bigger PR.
